Check that pytest-arraydiff is installed

astrofrog opened the pull request:
@cdeil - this is one way to make sure that pytest-arraydiff is installed and give a better error message. I'm not a fan of this, but it does the job:

```
$ python setup.py test
Traceback (most recent call last):
  File "/Users/tom/Dropbox/Code/Astropy/regions/regions/__init__.py", line 29, in <module>
    import pytest_arraydiff
ImportError: No module named 'pytest_arraydiff'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "setup.py", line 41, in <module>
    __import__(PACKAGENAME)
  File "/Users/tom/Dropbox/Code/Astropy/regions/regions/__init__.py", line 31, in <module>
    raise ImportError("The pytest-arraydiff package is required for the tests. "
ImportError: The pytest-arraydiff package is required for the tests. You can install it with: pip install pytest-arraydiff
```

I'm open to other ideas though.

bsipocz commented:
Now that a release is out, I'm adding ``pytest-arraydiff`` to the astropy conda channel, too. https://github.com/astropy/conda-channel-astropy/pull/102

cdeil commented:
There's this fail on circle-ci:
https://circleci.com/gh/astropy/regions/12?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

@astrofrog - Can you have a look?
Should `pytest-arraydiff` have an `install_requires` on pytest?

I don't really understand why that code is in the `else` block.
And doesn't raising a different `ImportError` lead to possible info lost when the original import failed because of some other problem (e.g. because numpy or pytest isn't there)?

---

How about this instead?

- Add `pytest-array-diff` to `extras_require/test` in setup.py [here](https://github.com/astropy/regions/blob/master/setup.py#L115)
- If that doesn't work properly, file a reminder issue that it shall be fixed later.
- Mention `pytest-array-diff` at http://astropy-regions.readthedocs.io/en/latest/installation.html#optional-dependencies and maybe also mention that devs should do `pip install regions[test]`.

Maybe I shouldn't even have said anything, it's just a few devs on this package for now, so a good error message isn't so important.

What will happen if users run `python -c 'import regions; regions.test()` to verify their installation and they don't have `pytest-array-diff`? Is it a required or optional dependency to run tests?

astrofrog commented:
More later - but a simpler solution is to just remove ``--arraydiff`` from ``setup.cfg`` which means the tests will run but just not compare the output (and pytest-arraydiff is then an optional dependency), then we specify the flag explicitly in the Travis config.

astrofrog commented:
@cdeil - I can try and finish this up

astrofrog commented:
@cdeil - ok, so here's what I've done:

* Moved the exception about pytest-arraydiff to ``conftest.py`` so it only gets considered when the tests get run.
* Added pytest-arraydiff to setup.py (removed pytest-mpl since it's not used?)
* Updated installation instructions to mention pytest-arraydiff but I didn't mention ``pip install regions[test]`` because people installing from PyPI aren't likely to be doing development. We *could* replace ``python setup.py install`` with ``pip install .[test]`` but not sure if that's really needed.

cdeil commented on regions/conftest.py:
```diff
@@ -2,6 +2,14 @@
+except ImportError:
+    raise ImportError("The pytest-arraydiff package is required for the tests. "
+                      "You can install it with: pip install pytest-arraydiff")
+else:
```
Is this else-block useful?
If the import fails, the symbol won't be in the namespace, i.e. the else block doesn't do anything useful, no?

astrofrog replied:
Pytest crashes if we don't remove the pytest_arraydiff variable from the namespace

cdeil replied:
OK. Can you leave that info as a one-line comment after `else:`?

cdeil commented on setup.py:
```diff
@@ -113,7 +113,7 @@
               'wcsaxes',
           ],
           test=[
-              'pytest-mpl',
```
I'd prefer we leave `pytest-mpl` here in `setup.py`.
We plan to use it very soon, so we added it to `setup.py` and `.travis.yml` already.

astrofrog replied:
Fixed

larrybradley commented:
:+1:
